editor/sidebar/post-taxonomies/hierarchical-term-selector.js

```diff
@@ -48,16 +58,87 @@ class HierarchicalTermSelector extends Component {
 		return fillWithChildren( termsByParent[ 0 ] || [] );
 	}
 
+	onChange( event ) {
+		const { onUpdateTerms, terms = [], restBase } = this.props;
+		const termId = parseInt( event.target.value, 10 );
+		const hasTerm = terms.indexOf( termId ) !== -1;
+		const newTerms = hasTerm
+			? without( terms, termId )
+			: [ ...terms, termId ];
+		onUpdateTerms( newTerms, restBase );
+	}
+
+	onChangeFormName( event ) {
+		this.setState( { formName: event.target.value } );
+	}
+
+	onChangeFormParent( event ) {
+		this.setState( { formParent: event.target.value } );
+	}
+
+	onToggleForm() {
+		this.setState( ( state ) => ( {
+			showForm: ! state.showForm,
+		} ) );
+	}
+
+	onAddTerm( event ) {
+		event.preventDefault();
+		const { formName, formParent } = this.state;
+		const findOrCreatePromise = new Promise( ( resolve, reject ) => {
+			this.setState( {
+				adding: true,
+			} );
+			// Tries to create a term or fetch it if it already exists
+			const Model = wp.api.getTaxonomyModel( this.props.slug );
+			this.addRequest = new Model( {
+				name: formName,
+				parent: formParent ? formParent : undefined,
+			} ).save();
+			this.addRequest
+				.then( resolve, ( xhr ) => {
+					const errorCode = xhr.responseJSON && xhr.responseJSON.code;
+					if ( errorCode === 'term_exists' ) {
```

Comment:
Any chance we can do a client-side validation against known terms to help bypass this?